js-pkg/packages/driftdb/src/webrtc.ts

```diff
@@ -0,0 +1,285 @@
+      }
+    }
+
+    return sendRec
+  }
+
+interface Syncable<T> extends Iterable<T> {
+  has(arg0: T): boolean
+}
+function* difference<T>(setA: Iterable<T>, setB: Syncable<T>) {
+  for (const elem of setA) {
+    if (!setB.has(elem)) yield elem
+  }
+}
+
+const mapToSyncable = <K>(a: Map<K, any>): Syncable<K> => {
+  return {
+    [Symbol.iterator]: a.keys.bind(a),
+    has: a.has.bind(a)
+  }
+}
+
+export class WebRTCConnections {
+  public connMap = new Map<string, ChannelSendReceive>()
+  private onMessage = (msg: DataChannelMsg) => {
+    console.log('unhandled', msg)
+  }
+
+  constructor(private db: DbConnection, public myId: string, private throttleMs = 0) {}
+
+  addNewConnection(p2: string) {
+    const signalingChannel = new SignalingChannel(this.db, this.myId, p2)
+    const conn = signalingChannel.createWebRTCConnection([
+      getDataChannelCreator(this.myId, p2, (msg) => this.onMessage(msg))
+    ])[0]
+    this.connMap.set(p2, conn)
+  }
+
+  send = throttle((msg: string) => {
+    this.connMap.forEach((peer) => {
+      peer.send(msg)
+    })
+  }, this.throttleMs)
+
+  setOnMessage(func: OnMessage) {
+    this.onMessage = func
+    this.connMap.forEach((peer) => {
+      peer.onMessage = func
+    })
+  }
+
+  removeConnection(peer: string) {
+    this.connMap.delete(peer)
+  }
+
+  peers() {
+    return mapToSyncable(this.connMap)
+  }
+}
+
+function sync<T>(a: Syncable<T>, b: Syncable<T>, cb1: (arg: T) => void, cb2: (arg: T) => void) {
+  for (const elem of difference(a, b)) {
+    cb1(elem)
+  }
+  for (const elem of difference(b, a)) {
+    cb2(elem)
+  }
+}
+
+export class SyncedWebRTCConnections extends WebRTCConnections {
+  presence: PresenceListener<string>
+  #peersToLastMsg: Record<string, WrappedPresenceMessage<any>> = {}
+  constructor(db: DbConnection, id: string, throttle = 0) {
+    super(db, id, throttle)
+    this.presence = new PresenceListener<string>({
+      initialState: '',
+      db,
+      clientId: id,
+      callback: (msg) => {
+        let peers = Object.values(msg).map(({ value }) => value)
+        this.sync({ [Symbol.iterator]: peers.values.bind(peers), has: peers.includes.bind(peers) })
+      },
+      minPresenceInterval: throttle
+    })
+    this.refreshConnections()
+    this.presence.subscribe()
+    this.setOnMessage((_msg) => {})
+  }
+
+  setOnMessage(func: OnMessage) {
+    super.setOnMessage((msg) => {
+      func(msg)
+      this.#peersToLastMsg[msg.sender] = msg
+    })
+  }
+
+  refreshConnections() {
+    let newId = (typeof crypto !== 'undefined' ? crypto.randomUUID() : 'PLACEHOLDER').slice(0, 5)
+    this.myId = newId
+    this.presence.updateState(newId)
+  }
+
+  sync(newPeers: Syncable<string>) {
+    newPeers = new Set(newPeers) //since this is used twice it can't be an iterable
+    sync(
+      {
+        [Symbol.iterator]: () => Object.keys(this.#peersToLastMsg).values(),
+        has: Reflect.has.bind({}, this.#peersToLastMsg)
+      },
```

Comment:
I gotta say, I don't think this whole `Syncable` business is paying off. It makes the code a lot more complicated in several places, has weird edge cases you end up having to deal with (e.g. above, having to turn newPeers into a Set anyway because it can't be used twice as an iterable), and generally just adds a lot of stuff that gets in the way of understanding what's really happening. There wasn't really anything wrong with the way it was before, as I recall. Can we just go back to that?

Reply:
I changed it cause that allocated a bunch every time sync was called, which was every single usePresence onMessage

Reply:
but yea I guess I'm still allocating a set here, sigh, I'll change it back.